src/rdkafka_sticky_assignor.c

```diff
@@ -25,170 +25,3465 @@
+
+                        if (partitions)
+                                ; /* Avoid unused warning */
+
+                        return rd_false;
+                }
+        }
+
+        return rd_true;
+}
+#endif
+
+
+/**
+ * @brief Comparator to sort ascendingly by rd_map_elem_t object value as
+ *        topic partition list count, or by member id if the list count is
+ *        identical.
+ *        Used to sort sortedCurrentSubscriptions list.
+ *
+ * elem.key is the consumer member id string,
+ * elem.value is the partition list.
+ */
+static int sort_by_map_elem_val_toppar_list_cnt (const void *_a,
+                                                 const void *_b) {
+        const rd_map_elem_t *a = _a, *b = _b;
+        const rd_kafka_topic_partition_list_t *al = a->value, *bl = b->value;
+        int r = al->cnt - bl->cnt;
+        if (r)
+                return r;
+        return strcmp((const char *)a->key, (const char *)b->key);
+}
+
+
+/**
+ * @brief Assign partition to the most eligible consumer.
+ *
+ * The assignment should improve the overall balance of the partition
+ * assignments to consumers.
+ */
+static void
+assignPartition (const rd_kafka_topic_partition_t *partition,
+                 rd_list_t *sortedCurrentSubscriptions /*rd_map_elem_t*/,
+                 map_str_toppar_list_t *currentAssignment,
+                 map_str_toppar_list_t *consumer2AllPotentialPartitions,
+                 map_toppar_str_t *currentPartitionConsumer) {
+        const rd_map_elem_t *elem;
+        int i;
+
+        RD_LIST_FOREACH(elem, sortedCurrentSubscriptions, i) {
+                const char *consumer = (const char *)elem->key;
+                const rd_kafka_topic_partition_list_t *partitions;
+
+                partitions = RD_MAP_GET(consumer2AllPotentialPartitions,
+                                        consumer);
+                if (!rd_kafka_topic_partition_list_find(partitions,
+                                                        partition->topic,
+                                                        partition->partition))
+                        continue;
+
+                rd_kafka_topic_partition_list_add(
+                        RD_MAP_GET(currentAssignment, consumer),
+                        partition->topic, partition->partition);
+
+                RD_MAP_SET(currentPartitionConsumer,
+                           rd_kafka_topic_partition_copy(partition), consumer);
+
+                /* Re-sort sortedCurrentSubscriptions since this consumer's
+                 * assignment count has increased.
+                 * This is an O(N) operation since it is a single shuffle. */
+                rd_list_sort(sortedCurrentSubscriptions,
+                             sort_by_map_elem_val_toppar_list_cnt);
+                return;
+        }
+}
+
+/**
+ * @returns true if the partition has two or more potential consumers.
+ */
+static RD_INLINE rd_bool_t
+partitionCanParticipateInReassignment (
```

Comment:
`partitionCanParticipateInReassignment`: it's like you're writing java. I guess this was on purpose, and there's an argument for not modifying the names, but when translating between languages, i've noticed it's typical to change the naming to be idiomatic (same comment many other places).

Reply:
I AM writing Java!